curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -693,55 +693,74 @@ void TopologyManager::CreatePartitions(const CreatePartitionRequest *request,
-                   << "erro code = " << ret;
-        return ret;
-    }
-
-    Pool pool;
-    if (!topology_->GetPool(poolId, &pool)) {
-        LOG(ERROR) << "Get pool failed.";
-        return TopoStatusCode::TOPO_POOL_NOT_FOUND;
-    }
-
-    int replicaNum = pool.GetReplicaNum();
-    std::set<ZoneIdType> zones;
-    ret = topology_->ChooseZonesInPool(poolId, &zones, replicaNum);
-    if (TopoStatusCode::TOPO_OK != ret) {
-        LOG(ERROR) << "Select Zone failed when create partition"
-                   << "poolId = " << poolId << "erro code = " << ret;
-        return ret;
-    }
-
+    std::set<PoolIdType> unavailablePools;
     std::set<MetaServerIdType> metaServerIds;
     std::set<std::string> metaServerAddrs;
-    for (auto zoneId : zones) {
-        MetaServerIdType metaServerId;
-        ret = topology_->ChooseSingleMetaServerInZone(zoneId, &metaServerId);
+    int replicaNum;
+    do {
+        TopoStatusCode ret = topology_->ChooseSinglePoolRandom(&poolId,
+                                                    unavailablePools);
         if (TopoStatusCode::TOPO_OK != ret) {
-            LOG(ERROR) << "Select metaServer failed when create partition"
-                       << "zoneId = " << zoneId << "erro code = " << ret;
+            LOG(ERROR) << "Select Pool failed when create partition"
+                       << ", erro code = " << ret;
             return ret;
         }
-        MetaServer metaServer;
-        if (topology_->GetMetaServer(metaServerId, &metaServer)) {
-            metaServerAddrs.emplace(
-                metaServer.GetInternalHostIp() + ":" +
-                std::to_string(metaServer.GetInternalPort()));
-        } else {
-            LOG(ERROR) << "Get metaserver failed.";
-            return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
+
+        Pool pool;
+        if (!topology_->GetPool(poolId, &pool)) {
+            LOG(ERROR) << "Get pool failed.";
+            return TopoStatusCode::TOPO_POOL_NOT_FOUND;
         }
-        metaServerIds.emplace(metaServerId);
-    }
 
-    if (metaServerAddrs.size() != replicaNum) {
-        LOG(ERROR) << "Have invalid metaserver number when create copyset."
-                   << " metaserver number = " << metaServerAddrs.size()
-                   << ", replicanum = " << replicaNum;
-    }
+        replicaNum = pool.GetReplicaNum();
+        int needZoneNum = replicaNum;
+        std::set<ZoneIdType> zones;
+        std::set<ZoneIdType> unavailableZones;
+        do {
+            ret = topology_->ChooseZonesInPool(poolId, &zones, unavailableZones,
+                                            needZoneNum);
+            if (TopoStatusCode::TOPO_OK != ret) {
+                LOG(ERROR) << "Select Zone failed when create partition"
+                        << "poolId = " << poolId << ", erro code = " << ret;
+                unavailablePools.emplace(poolId);
+                metaServerIds.clear();
+                metaServerIds.clear();
```

Comment:
clear twice, and I think you should not clear here, copyset must create in the same pool, but different replicas can create in any zone.

Reply:
> clear twice, and I think you should not clear here, copyset must create in the same pool, but different replicas can create in any zone.

Yes, here is have no available zones in this pool, need choose another pool